### DISCUSSION TITLE
How to make groups

### DISCUSSION THREAD

**Discussion by HSDSZ** [bnpr/Malt#98] [Q&A] [answered]
First, thank you so much for making this Malt engine. I find myself getting more used to code based texture now. HAHA

I have one question.
After I cretae some varibles, Can I group them?
For example, in the bottom pic
Can I group the "Base Color" and "Texture" into a New group like the existing "Lights Groups"?

![questiion](https://user-images.githubusercontent.com/36077492/125198696-9b64a180-e295-11eb-86e5-22a7c9b0e10b.png)

**Comment by pragma37** [comment, marked as answer]
Hi! I'm glad that Malt is helping you to get into shader coding. :)

For grouping variables, you can use structs. For example:

```GLSL
#include "Pipelines/NPR_Pipeline.glsl"

struct TextureSettings
{
    sampler2D texture;
    int uv_index;
};

uniform TextureSettings Color_Texture;

void COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO)
{
    PO.color = texture(Color_Texture.texture, UV[Color_Texture.uv_index]);
}
```

Alternatively, you can also use `_0_` as a separator in the variable name.

```GLSL
#include "Pipelines/NPR_Pipeline.glsl"

uniform sampler2D Color_Texture_0_texture;
uniform int Color_Texture_0_uv_index;

void COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO)
{
    PO.color = texture(Color_Texture_0_texture, UV[Color_Texture_0_uv_index]);
}
```

Both of these samples will look the same in the UI.

**Reply by HSDSZ** [reply]
Many thanks. It works

**Reply by HSDSZ** [reply]
How Can I initialize Sampler2D in a struct. Or can I skip the Sampler2D initialization?

The following code works if I remove sampler2D in the struct. But with Sampler2D. I don;t know how to make it work

struct ImageSettings
{
    bool Enbale_Image;
    int Image_UV_Index;
    vec2 Scale;
    vec2 Offset;
    sampler2D Image;
};
uniform ImageSettings imagesettings = ImageSettings(false, 0, vec2(1.0,1.0), vec2(0.0, 0.0));

**Reply by pragma37** [reply]
Looks like I forgot to answer your second question, sorry!
Sadly, there's no way to initialize sampler2D variables from GLSL.

**Reply by HSDSZ** [reply]
HAHA, thank you anyway

**Comment by HSDSZ** [comment]
May I ask, How to make group with node?

**Reply by pragma37** [reply]
There's no node groups support at the moment.
You could rewrite your nodes as a function and put it into your Shader Library.
https://youtu.be/tE99jgCCcNE?t=706
